algorithms/src/crh/bhp.rs

```diff
@@ -96,164 +111,34 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
-                                    let mut encoded = *g;
-                                    if (i & 0x01) != 0 {
-                                        encoded += g;
-                                    }
-                                    if (i & 0x02) != 0 {
-                                        encoded += g.double();
-                                    }
-                                    if (i & 0x04) != 0 {
-                                        encoded = encoded.neg();
-                                    }
-                                    *element = encoded;
-                                }
-                                out
-                            })
-                            .collect()
-                    })
-                    .collect())
-            })
-            .expect("failed to init BoweHopwoodPedersenCRHParameters")
-    }
-
-    /// Precondition: number of elements in `input` == `num_bits`.
-    pub(crate) fn hash_bits_inner<S: Borrow<bool>>(
-        &self,
-        input: impl Iterator<Item = S>,
-        num_bits: usize,
-    ) -> Result<G, CRHError> {
-        if num_bits > WINDOW_SIZE * NUM_WINDOWS {
-            return Err(CRHError::IncorrectInputLength(num_bits, WINDOW_SIZE, NUM_WINDOWS));
+    pub(crate) fn hash_bits_inner(&self, input: &[bool]) -> Result<G, CRHError> {
+        // Ensure the input size is within the parameter size,
+        if input.len() > NUM_WINDOWS * WINDOW_SIZE * BHP_CHUNK_SIZE {
+            return Err(CRHError::IncorrectInputLength(input.len(), WINDOW_SIZE, NUM_WINDOWS * BHP_CHUNK_SIZE));
         }
-        debug_assert!(WINDOW_SIZE <= MAX_WINDOW_SIZE);
-        debug_assert!(NUM_WINDOWS <= MAX_NUM_WINDOWS);
 
-        // overzealous but stack allocation
-        let mut buf_slice = [false; MAX_WINDOW_SIZE * MAX_NUM_WINDOWS + BOWE_HOPWOOD_CHUNK_SIZE + 1];
-        buf_slice[..num_bits].iter_mut().zip(input).for_each(|(b, i)| *b = *i.borrow());
-
-        let mut bit_len = WINDOW_SIZE * NUM_WINDOWS;
-        if bit_len % BOWE_HOPWOOD_CHUNK_SIZE != 0 {
-            bit_len += BOWE_HOPWOOD_CHUNK_SIZE - (bit_len % BOWE_HOPWOOD_CHUNK_SIZE);
-        }
-
-        debug_assert_eq!(bit_len % BOWE_HOPWOOD_CHUNK_SIZE, 0);
-
-        debug_assert_eq!(
-            self.bases.len(),
-            NUM_WINDOWS,
-            "Incorrect number of windows ({:?}) for BHP of {:?}x{:?}x{}",
-            self.bases.len(),
-            WINDOW_SIZE,
-            NUM_WINDOWS,
-            BOWE_HOPWOOD_CHUNK_SIZE,
-        );
-        for bases in self.bases.iter() {
-            debug_assert_eq!(bases.len(), WINDOW_SIZE);
-        }
-        let base_lookup = self.base_lookup(&self.bases);
-        debug_assert_eq!(base_lookup.len(), NUM_WINDOWS);
-        for bases in base_lookup.iter() {
-            debug_assert_eq!(bases.len(), WINDOW_SIZE);
+        // Pad the input to a multiple of `BHP_CHUNK_SIZE` for hashing.
+        let mut input = input.to_vec();
```

Comment:
We might want to keep the stack allocation logic

Reply:
The stack allocation logic requires a worst case allocation, which ends up being too expensive.